Feature: fungible tokens on account

luixo opened the pull request:
There is no design here, but you can check out the data flow on page:

https://frontend-pr-969-kgmz.onrender.com/beta/accounts/v2.ref-finance.near
https://frontend-pr-969-kgmz.onrender.com/beta/accounts/vlad.near/fungible-tokens/kongztoken.near

shelegdmitriy commented on backend/src/db-utils.ts:
```diff
@@ -748,6 +748,44 @@ const queryAccountsList = async (
+      }
+    >(
+      [
+        `SELECT DISTINCT emitted_by_contract_account_id as "contractId"
```
```suggestion
        `SELECT DISTINCT emitted_by_contract_account_id AS "contractId"
```

shelegdmitriy reviewed (changes requested):
LGTM. But we need some design here. May be we can display it as a table (like on validators page,  but with three columns)
| Token | Author                                              | Balance                            |
| Utopia | utopia.secretskelliessociety.near | 373624.61642350 UTO |
as a temporary solution and not to procrastinate?

luixo commented:
> LGTM. But we need some design here. May be we can display it as a table (like on validators page, but with three columns)
| Token | Author | Balance |
| Utopia | utopia.secretskelliessociety.near | 373624.61642350 UTO |
as a temporary solution and not to procrastinate?

Do we need to rush actually?
We can add USN as a promo FT for example, but for a proper release we need proper design

luixo commented:
This one is ready for review.
It lacks some of the features (like "new balance") at the moment.

frol commented:
* [ ] NEAR native coin is missing. It should use a different source, but the same representation
* [ ] Sync with @reefoh on how we want the balances displayed (e.g. should we display at least 2 decimals after the floating point ["3.00" instead of just "3"], or should we add the symbol after the value like "3 USN" or "3.00 USN" or "3 $USN")
* [ ] Icons of some of the coins are missing (USN and KONGZ don't have the icon: https://frontend-pr-969-kgmz.onrender.com/beta/accounts/vlad.near/fungible-tokens/usn)
* [ ] In order to pull the new balance, make `ft_balance_of()` view call to the contract on the block height of the last asset event you pull, and then apply the deltas of the balance change to all the newer records

P.S. making view-calls through near-api-js is still convoluted, and does not let you specify the block height ([it has hardcoded `finality: "final"`](https://github.com/near/near-api-js/blob/3206032df37a3d7fadd7c9a8325711ec9a629293/src/account.ts#L540-L568)), so I recommend we further extend [our implementation in near-explorer](https://github.com/near/near-explorer/blob/9e032fb8e182db18e962c16db4c8168edd2241cc/backend/src/utils/near.ts#L30) with the following implementation that adds `blockReference` (I tried to be as realistic as possible, but I have not tested this code):

```ts
async viewFunction(
        contractId: string,
        methodName: string,
        args: any = {},
        { parse = parseJsonFromRawResponse, stringify = bytesJsonStringify, jsContract=false, blockReference = {finality: "final"} } = {}
    ): Promise<any> {
        let encodedArgs;
        this.validateArgs(args);
    
        if(jsContract){
            encodedArgs = this.encodeJSContractArgs(contractId, methodName, Object.keys(args).length >  0 ? JSON.stringify(args): '');
        } else{
            encodedArgs =  stringify(args);
        }

        const result = await this.connection.provider.query<CodeResult>({
            request_type: 'call_function',
            account_id: jsContract ? this.connection.jsvmAccountId : contractId,
            method_name: jsContract ? 'view_js_contract'  : methodName,
            args_base64: encodedArgs.toString('base64'),
            ...blockReference,
        });

        if (result.logs) {
            this.printLogs(contractId, result.logs);
        }

        return result.result && result.result.length > 0 && parse(Buffer.from(result.result));
    }
```

And then you will be able to pass `blockReference: { block_id: 68217757 }`

frol commented on frontend/src/components/beta/accounts/AccountFungibleTokenHistory.tsx:
```diff
@@ -0,0 +1,169 @@
+          disabled={!item.counterparty}
+        >
+          <TableElement>
+            {item.counterparty
+              ? shortenString(item.counterparty)
```
Why don't we use AccountLink component? It makes perfect sense to be able to click on the account to navigate to that account details page.

luixo replied:
Well, actually, the thing is that I'll have to modify it quite a lot to use:
- we probably need the whole div around the name be clickable
- I'll have to work around disabled option for mint / burn events
- we need to derive if it's beta to have a proper link

Let's unify account link in another PR.

frol commented on frontend/src/components/beta/accounts/AccountFungibleTokenHistory.tsx:
```diff
@@ -0,0 +1,169 @@
+        <TableElement>
+          <TokenAmount
+            token={{ ...token, balance: item.amount }}
+            sign={item.direction === "out" ? -1 : 1}
```
This looks odd to me to use -1 / 1 as part of the component interface, to be honest. Why don't just use `balance` sign for that?

luixo replied:
I think we can add a check for negative numbers in `formatToPowerOfTen`, that will work.
We still need some specific prop to show "+" there, I chose a "prefix".

frol commented on frontend/src/components/beta/accounts/AccountFungibleTokenHistory.tsx:
```diff
@@ -0,0 +1,169 @@
+      "account-fungible-token-history",
+      { accountId, tokenAuthorAccountId: token.authorAccountId },
+    ]);
+    if (tokenHistoryQuery.status !== "success") {
+      if (tokenHistoryQuery.status === "loading") {
```
Somehow I felt like it is more idiomatic to use UPPER_SNAKE_CASE for the string consts.

luixo replied:
That's `react-query` status constant.

frol commented on frontend/src/components/utils/TokenAmount.tsx:
```diff
@@ -0,0 +1,54 @@
+      }
+    >
+      <span>
+        {props.sign === -1 ? "-" : ""}
```
We should display `+` when it is not `-` if we display balance *change*

<img width="554" alt="image" src="https://user-images.githubusercontent.com/304265/174879741-c3413a51-a93d-4d81-ad35-62a8d43310aa.png">

frol commented on frontend/src/pages/beta/accounts/[...slug].tsx:
```diff
@@ -82,18 +90,27 @@ const AccountQueryView: React.FC<QueryProps> = React.memo((props) => {
 export const getServerSideProps: GetServerSideProps<
   {},
-  { id: string }
+  { slug: string[] }
```
Do we have to give up on Next.js routing to avoid the account header re-render when user navigates between the tabs? It seems that recently added [Nested Layouts feature in Next.js](https://nextjs.org/blog/layouts-rfc) should help here

shelegdmitriy commented:
When I cheked this [preview link](https://frontend-pr-969-kgmz.onrender.com/beta/accounts/v2.ref-finance.near) I saw some strange things:

1. The first thing you see is the list of FTs on the left, and a lot of blank space on the right:<img width="813" alt="Снимок экрана 2022-06-22 в 22 41 08" src="https://user-images.githubusercontent.com/34593263/175122593-61967360-15b3-4193-ab3a-10b121f7a4ba.png">
Maybe it's only weird for me but maybe the first item of FTs list active by default?
2. When you choose the list item with a huge history you have to wait some time (quite a lot of time) (try to select *Utopia* from [preview link](https://frontend-pr-969-kgmz.onrender.com/beta/accounts/v2.ref-finance.near)). Also the list of history is too long to scroll it down. Maybe we can use pagination to avoid all it?

Overall it looks good, and all things I mention could solve later.

luixo commented:
> NEAR native coin is missing. It should use a different source, but the same representation

I guess we eventually decided to keep tokens without NEAR token for now, but we probably need to discuss how can we help with expectation of NEAR token history being there.

> Sync with @reefoh on how we want the balances displayed (e.g. should we display at least 2 decimals after the floating point ["3.00" instead of just "3"], or should we add the symbol after the value like "3 USN" or "3.00 USN" or "3 $USN")

I'll discuss the decimals. About the symbols - that's what we do now - display the symbols. We will not do "$USN" as a generic decision definitely - because `usn` contract has symbol `USN` in it, not `$USN`. We can apply some custom formatting for known FTs though.

> Icons of some of the coins are missing (USN and KONGZ don't have the icon: https://frontend-pr-969-kgmz.onrender.com/beta/accounts/vlad.near/fungible-tokens/usn)

I made a base64 validation function, but further research gave me an idea that react protects us from plain XSS attacks and browser protects us from CSRF attacks. Unfortunately, we'll have to add image sources manually in `next.config.js` (like for KONGZ logo)

> In order to pull the new balance, make ft_balance_of() view call to the contract on the block height of the last asset event you pull, and then apply the deltas of the balance change to all the newer records

That's a neat idea, actually!

luixo commented:
> In order to pull the new balance, make ft_balance_of() view call to the contract on the block height of the last asset event you pull, and then apply the deltas of the balance change to all the newer records

Actually, as we don't have pagination yet, we can just use zero as initial balance.

luixo commented:
> Maybe we can use pagination to avoid all it?

I made a hard limit 200 for now, we'll implement it in the next PR.

frol commented:
> It would be a little tough as we don't know what's the first FT for the user at page fetch and we keep path as our source of truth.

We could use /coins as our main URL, and then redirect to the first FT token when the list is known. Having NEAR native token on the list will, however simplify it, so we might as well just skip this auto-selection implementation now in favor of using /coins/NEAR in the future.

> I made a hard limit 200 for now, we'll implement it in the next PR.

👍 

> Actually, as we don't have pagination yet, we can just use zero as initial balance.

Given the above and also given the JSON Events support was not always there, we should not use 0 as our initial balance.

frol reviewed (changes requested):
Prefix in the tooltip is off:

<img width="395" alt="image" src="https://user-images.githubusercontent.com/304265/175770783-75f49142-01a4-4f81-86ef-a66893058395.png">

New balance sometimes is odd:

<img width="1713" alt="Screen Shot 2022-06-25 at 13 02 06" src="https://user-images.githubusercontent.com/304265/175770801-83175abe-b06c-4f49-8be2-0355ca56eec9.png">

Overall, I feel this PR is in a good shape to be merged and iterate on the design side in the next PRs.

luixo commented:
> We could use /coins as our main URL, and then redirect to the first FT token when the list is known. Having NEAR native token on the list will, however simplify it, so we might as well just skip this auto-selection implementation now in favor of using /coins/NEAR in the future.

I'm not sure we should use `coins` at all as not all fungible tokens are coins.

> Given the above and also given the JSON Events support was not always there, we should not use 0 as our initial balance.

Yep, found it the hard way :)

> Prefix in the tooltip is off:

Fixed!

> New balance sometimes is odd:

Fixed!
